corehq/apps/locations/views.py

```diff
@@ -118,23 +120,43 @@ def _get_loctype_info(self, loctype):
     def post(self, request, *args, **kwargs):
         payload = json.loads(request.POST.get('json'))
 
-        def mk_loctype(loctype):
-            loctype['allowed_parents'] = [p or '' for p in loctype['allowed_parents']]
-            cleaned_code = unicode_slug(loctype['code'])
-            if cleaned_code != loctype['code']:
-                err = _(
-                    'Location type code "{code}" is invalid. No spaces or special characters are allowed. '
-                    'It has been replaced with "{new_code}".'
+        def mk_loctype(name, code, allowed_parents, administrative,
+                       shares_cases, view_descendants):
+            parents = allowed_parents
+            if len(parents) > 1:
+                messages.warning(request, _("Location types cannot have "
+                                            "multiple parent types, using {}"
+                                            .format(parents[0])))
+            if parents and parents[0]:
+                parent, __ = LocationType.objects.get_or_create(
+                    domain=self.domain,
+                    name=parents[0],
                 )
-                messages.warning(request, err.format(code=loctype['code'], new_code=cleaned_code))
-                loctype['code'] = cleaned_code
-            return LocationType(**loctype)
+            else:
+                parent = None
 
-        #TODO add server-side input validation here (currently validated on client)
+            location_type, __ = LocationType.objects.get_or_create(
+                domain=self.domain,
+                name=name,
+            )
 
-        self.domain_object.location_types = [mk_loctype(l) for l in payload['loc_types']]
+            cleaned_code = unicode_slug(code)
+            if cleaned_code != code:
+                err = _('Location type code "{code}" is invalid. No spaces or '
+                        'special characters are allowed. It has been replaced '
+                        'with "{new_code}".')
```

Comment:
I think the name is misleading, looks like unicode_slug accepts unicode, but the output is ascii (it calls out to [this django function](https://github.com/django/django/blob/master/django/utils/text.py#L413-L423))